Design document for Post Reconciliation Extension (#2812)
Azure/azure-service-operator: docs/hugo/content/ADR-2023-03-Post-Reconciliation-Extension.md

---
title: '2023-03: Post Reconciliation Extensions'
---

## Context

The current behaviour of Azure Service Operator (ASO) is to set the `Ready` status on a resource by default after a successful creation of a resource. 

> For some resources, this is not ideal as we need to check the status of a created resource post reconciliation and set a condition accordingly. 

Reviewer: Remember the extension point is generic - while our current use is for `PrivateEndpoint`, we're not designing it to handle that case only. The design and the description need to allow for many uses - the defining characteristic of this extension point is _when_ it gets applied, not _which_ resources it's applied to.

```suggestion
For some resources, this is not sufficient as there may be additional steps required after initial creation of the resource. We need the ability to check the status of a created resource post reconciliation and set a condition accordingly. 
```

An example of this:

> - `PrivateEndpoints` with `ConnectionState` of `Approved`, `Rejected` or `Pending`.

Reviewer: If there's one example, put it in  sentence. Only use a bullet point if there's more than one.

Author: Done!

> We need a way to allow the operator to handle the above case in a way where we can write an extension point to be called post reconciliation. Which allows the operator to set a custom condition on the resource based on the status we've received from ARM.

Reviewer: ```suggestion
We need a way to extend the operator to handle the above case, an extension point to be called post reconciliation allowing handwritten code to set a custom condition on the resource based on the status we've received from ARM.
```

## Decision

We'll define a new extension called `PostReconciliationChecker` in similar way to our `PreReconciliationChecker` extension point. 

The extension point will receive the following parameters:

- The current resource, with a status freshly updated from Azure.
- A Kubernetes client allowing for Cluster operations.
- An ARM client allowing for ARM operations.
- A logger to allow for tracing of what the extension did.
- A context to allow cancellation of long-running operations.

The above would return one of two below options:

- `Success` if the post reconcile check is met.
- `Failure` with `NewReadyConditionImpactingError` if the post reconcile check is not met.

> 

Reviewer: What does the return value mean?

If an implementation of this extension returns `Success`, what happens? How is that return value structured?
If it returns `Failure`, what does that mean for the reconciler? 
How does the `NewReadyConditionImpactingError` get returned? 
Is it possible for other errors to be returned? How?

Author: Have elaborated more on this.
